### PR TITLE
V5

### PR DESCRIPTION
**Pull request by rightjelkin**
New cli commands:
- deployment upgrade (creates a new version of deployment by deployment name)
- deployment destroy (destroy given deployment version by deployment and version names)
- deployment version list (returns list of deployment versions by name)
Updated:
- deployment list (new field from versioning logic)

### CODE REVIEW

**Comment by rightjelkin** [comment]
Forgot about unit-tests fixing according on new API

File: packages/hydra-cli/src/commands/deploy.ts [hunk 1]
```diff
@@ -28,15 +28,51 @@ export default class Deploy extends Command {
...
+        `Head origin commit is not the same as the local origin commit or there are unstaged or uncomitted changes`
+      )
+      const enterResult = await cli.prompt(
+        `Are you sure (enter corresponding remote number): \n1.yes\n[any other key].no\n`,
+        { required: true }
+      )
+      if (enterResult !== '1') {
+        this.log('Canceled')
+        return
+      }
```
Comment (dzhelezov) [inline packages/hydra-cli/src/commands/deploy.ts:70]: Let's always show an info line: `Releasing the Squid at ${remoteUrl}` before the actual deployment starts

File: packages/hydra-cli/src/commands/deployment-versions/ls.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+export default class Ls extends Command {
+  static description = 'Deployment version list'
+
+  static flags = {
+    'deployment-name': flags.string({
+      description: 'deployment name',
+      required: true,
+    }),
+    'no-truncate': flags.boolean({
```
Comment (eldargab) [inline packages/hydra-cli/src/commands/deployment-versions/ls.ts:16]: If it is true by default, it should be named `truncate` a be false by default.

File: packages/hydra-cli/src/commands/deployments/ls.ts [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
 import { cli } from 'cli-ux'
 import { deploymentList } from '../../rest-client/routes/deployments'
+import Debug from 'debug'
 
+const debug = Debug('qnode-cli:deployment-list')
 export default class Ls extends Command {
   static description = 'Deployments list'
 
+  static flags = {
+    'no-truncate': flags.boolean({
```
Comment (eldargab) [inline packages/hydra-cli/src/commands/deployments/ls.ts:11]: Should be just `truncate`

File: packages/hydra-cli/src/commands/deployment-versions/ls.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { deploymentVersionList } from '../../rest-client/routes/deploymentVersions'
+import Debug from 'debug'
+
+const debug = Debug('qnode-cli:deployment-version-list')
+
+export default class Ls extends Command {
+  static description = 'Deployment version list'
+
+  static flags = {
+    'deployment-name': flags.string({
```
Comment (eldargab) [inline packages/hydra-cli/src/commands/deployment-versions/ls.ts:12]: There should be just a one command for listing of all deployments. For deployment identification it should output 2 columns: app name and version. To output versions of just a one app, optional flag can be used.

File: packages/hydra-cli/src/commands/upgrade.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  async run(): Promise<void> {
+    const { flags } = this.parse(Upgrade)
+    debug(`Parsed flags: ${JSON.stringify(flags, null, 2)}`)
+    const deploymentName = flags.name
+
+    const remotes = await git.getRemotes(true)
+    const remoteUrl = remotes.find((remote) => remote.name === remoteUrlName)
+    if (!remoteUrl) {
+      throw Error(`Remote url with name ${remoteUrlName} not exists`)
```
Comment (eldargab) [inline packages/hydra-cli/src/commands/upgrade.ts:33]: rude

File: packages/hydra-cli/src/commands/deploy.ts [hunk 6]
```diff
@@ -28,15 +28,51 @@ export default class Deploy extends Command {
...
 
     const remotes = await git.getRemotes(true)
-    const remoteUrl = remotes.find((remote) => remote.name === remoteUrlName)
-    if (!remoteUrl) {
-      throw Error(`Remote url with name ${remoteUrlName} not exists`)
+    let remotesCLIList = ''
+    for (let i = 1; i <= remotes.length; i++) {
+      remotesCLIList += `\n${i}. ${remotes[i - 1].name}`
+    }
+    const remoteNumber = await cli.prompt(
```
Comment (eldargab) [inline packages/hydra-cli/src/commands/deploy.ts:35]: May be use library like https://github.com/cyrilwanner/cli-select

File: packages/hydra-cli/src/commands/deploy.ts [hunk 7]
```diff
@@ -28,15 +28,51 @@ export default class Deploy extends Command {
...
+    const remoteUrl = remotes[Number(remoteNumber) - 1]
+    if (remoteUrl === undefined) {
+      throw Error(`Incorrect input`)
+    }
+    const currentBranch = (await git.branch()).current
+
+    // show warning if head origin commit is not the same as the
+    // local origin commit or there are unstaged or uncomitted changes
+    const status = await git.status()
+    const diffs = await git.diffSummary([
```
Comment (eldargab) [inline packages/hydra-cli/src/commands/deploy.ts:48]: This is incorrect. First of all, remote can be stale and we can get false equality. Second, there is no need for such a clever check here. Branches are equal, when their commit hashes are equal.

File: packages/hydra-cli/src/commands/deploy.ts [hunk 8]
```diff
@@ -28,15 +28,51 @@ export default class Deploy extends Command {
...
+      this.warn(
+        `Head origin commit is not the same as the local origin commit or there are unstaged or uncomitted changes`
+      )
+      const enterResult = await cli.prompt(
+        `Are you sure (enter corresponding remote number): \n1.yes\n[any other key].no\n`,
+        { required: true }
+      )
+      if (enterResult !== '1') {
+        this.log('Canceled')
+        return
```
Comment (eldargab) [inline packages/hydra-cli/src/commands/deploy.ts:69]: Exit code should be non-zero

**Review by eldargab — CHANGES_REQUESTED** [review]
Apart from inline comments, command names are clearly ugly.

My suggestion would be:

```
app:deploy
app:delete
app:switch
app:ls
```

**Comment by eldargab** [comment]
May be we can define the target remote branch as an upstream of a current local branch?

**Pull request merged**